core/dnsserver/server.go

```diff
@@ -155,6 +155,10 @@ func (s *Server) Address() string { return s.Addr }
 // defined in the request so that the correct zone
 // (configuration and middleware stack) will handle the request.
 func (s *Server) ServeDNS(w dns.ResponseWriter, r *dns.Msg) {
+	s.ServeDNSWithContext(context.Background(), w, r)
+}
+
+func (s *Server) ServeDNSWithContext(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) {
```

Comment:
We could get a little crazy with this to, to help track down issues in fall through chains like described here https://github.com/coredns/coredns/issues/515. You could bundle the necessary trace data into an EDNS0 local option so you could actually trace through DNS calls not just gRPC.